### ISSUE TITLE
After Tag-Grouping APIs, SwaggerUI renders empty controllers/empty tags

### ISSUE THREAD

**Issue by taddygrrr** [swagger-api/swagger-ui#4157]
After upgrading from a 3.x version to a version of Swagger-UI that provides the schemes in the UI, I noticed that my controllers started rendering in the UI above the Tag-Grouped APIs.   Since all the APIs are grouped by tags, none of them are in the controllers.    Rendering empty controllers.

@ApiIgnore remove all the APIs in a controller from the Tag-Group.

| Q                               | A
| ------------------------------- | -------
| Bug or feature request?         |     Bug
| Which Swagger/OpenAPI version?  | 2.0
| Which Swagger-UI version?        |    3.1.x
| How did you install Swagger-UI? |    Download
| Which browser & version?        | Chrome, Safari, Firefox  
| Which operating system?         |  OSX

## Model from which spec is generated
@RestController
public class SomeController {

@ApiOperation(value="Operation Name", tags="Query Interfaces", notes="Operation Note")
@GetMapping(value = "/operation", produces = { APPLICATION_JSON })
@ResponseBody
ResponseEntity<SomeResponse> getResponse(
   @ApiParam(value="someId", required = false) @RequestParam(value = "someId", required=false) String someId) 
{
    return new ResponseEntity<SomeResponse>(...);
}

### Expected Behavior
Don't render controllers with no APIs in them.   "Query Interfaces" should show up with GET /operation, but empty SomeController should not.

### Current Behavior
Rendering empty controllers.

### Possible Solution
Don't render empty controllers, or make it an option.

### Context
Hide empty controllers.

**Comment by shockey** [comment]
Hi @taddygrrr, I'm not sure what to do with the model you've provided (I'm not a Java dev :smile:). Can you provide a YAML OpenAPI document that demonstrates your issue?

Thanks!

**Comment by taddygrrr** [comment]
Having looked at what's being generated for the spec (spec is generated by springfox, rendered by static distribution of swagger-ui), tried it out in swagger-ui editor,  I see what's going on.   At the root, a tag is generated for the controller.  This tag is not associated with any API but is still rendered.   I guess you could argue that the problem is with the spec.    I'm not sure how to prevent generation of the 'stuff-controller' tag below.    I'd think it shouldn't be rendered as in previous versions of swagger-ui.   Spec follows:



swagger: '2.0'
info:
  description: 'An API to publish, query, modify and delete stuff.'
  version: v1.0.0.0
  title: Some Title
host: localhost
basePath: /
tags:
  - name: stuff-controller
    description: Stuff Controller
schemes:
  - http
  - https
paths:
  /operation:
    post:
      tags:
        - Stuff Update Interfaces
      summary: Report new stuff or update existing stuff
      description: Create or update some stuff.
      operationId: operationUsingPOST
      consumes:
        - application/json;charset=UTF-8
      produces:
        - '*/*'
      parameters:
        - name: Authorization
          in: header
          description: Header containing user account id.
          required: false
          type: string
        - in: body
          name: stuffRequest
          description: This is a description of some stuff
          required: true
          schema:
            $ref: '#/definitions/StuffRequest'
      responses:
        '200':
          description: OK
          schema:
            type: object
definitions:
  StuffRequest:
    type: object
    properties:
      age:
        type: number
        format: float
        description: Age of stuff.

**Comment by taddygrrr** [comment]
Is there a way to remove the "waiting for feedback" label since I've given feedback?   If so, I'm not seeing it.

**Comment by heldersepu** [comment]
I think the problem is in the [core/plugins/spec/selectors.js](https://github.com/swagger-api/swagger-ui/blob/7af7121e64178fe4f3363c6e6c21b0afe4623f3b/src/core/plugins/spec/selectors.js#L214)

**Comment by shockey** [comment]
Hi @taddygrrr! You'd need write access to remove the label - I've removed it now. Sorry for the delay!

Rendering all tags defined in the definition is intended behavior. We tend to decide to render what's closest to the raw content of an API definition, rather than try to be clever about what the user intended - see https://github.com/swagger-api/swagger-ui/issues/4166 for a recent example of such a decision.

Someone did suggest that unused tags be rendered differently (https://github.com/swagger-api/swagger-ui/issues/3819), for example without an expand/collapse button, which sounds like a good indicator that the tag doesn't contain anything.

---

If you'd like to hide empty tags, here's a plugin that will do that for you!

```js
  const HideEmptyTagsPlugin = () => {
    return {
      statePlugins: {
        spec: {
          wrapSelectors: {
            taggedOperations: (ori) => (...args) => {
              return ori(...args)
                .filter(tagMeta => tagMeta.get("operations") && tagMeta.get("operations").size > 0)
            }
          }
        }
      }
    }
  }
```

**Comment by heldersepu** [comment]
Glad to see @shockey comment I was already working on a PR to hide the orphan tags...

**Comment by heldersepu** [comment]
Now that I read the comments from 3819 closely: https://github.com/swagger-api/swagger-ui/issues/3819#issuecomment-339424843
One would think that empty tags should not be there at all

**Comment by shockey** [comment]
> One would think that empty tags should not be there at all

You mean it shouldn't be in the document at all?

**Comment by heldersepu** [comment]
@shockey Yes, I think if there is the need to hide them, there should not be on the swagger.json doc
Even the @taddygrrr acknowledge that fact
> Though you could argue that the problem is with the spec,

but he counters with:
> shouldn't be rendered as in previous versions of swagger-ui

I guess we can say the  bug was on the previous version not showing the tags, and close this issue.

**Comment by shockey** [comment]
> Yes, I think if there is the need to hide them, there should not be on the swagger.json doc

Yep. There's no rule _against_ adding unused tags to your document, but there's also nothing in the spec that would suggest that unused tags would be ignored by display tools.

> I guess we can say the bug was on the previous version not showing the tags, and close this issue.

Seems like a fair call to me. As for the issue state, I want to keep this open for a bit to give @taddygrrr a chance to weigh in 😄

**Comment by taddygrrr** [comment]
Now that I fully understand what's going on, I'd say it's not a bug but a default behavior; I prefer the previous default.   It's hard for me to understand why someone would want empty tags rendered.   I think a better discussion of that would be under the feature issue request #3819.   Since this is referenced in #3819, I'm fine with closing it, perhaps changing it to feature request.  I'll add a comment there.  Thanks a bunch for considering the issue.

**Comment by shockey** [comment]
Agreed, closing in favor of https://github.com/swagger-api/swagger-ui/issues/3819.

Thanks for your contributions, @taddygrrr! Even the issues we don't end up taking action on help us understand how folks are using the tool 😄

**Comment by Tintifax79** [comment]
Thanks @shockey for the hint with the plugin. I'm facing a similar issue with empty Controller paths now with Swagger UI and Codegen as Maven Plugins in a Spring Boot project, and I'm wondering how I can inject the plugin you've posted in such an environment?

**Comment by shockey** [comment]
@Tintifax79, can you show me how you're embedding Swagger UI in your project?

**Comment by Tintifax79** [comment]
Oh, I'm very sorry I forgot about the thread. Meanwhile I have figured out how to resolve the issue (not knowing 100% was has caused it, so I can only make an educated guess):

With version 2.7.0 of the Maven Swagger UI plugin I had to repeat the function parameter annotations (@ApiParam, @RequestParam) in the definition of my functions for my @Controller class, which implemented the auto-generated Api interface (generated by Swagger Codegen). Seems like version 2.7.0 was only looking at the controller class, not at the parent/super class.

With version 2.9.2 (or maybe even with an earlier version between 2.7.0 and 2.9.2) of the Maven Swagger UI plugin there's no need to repeat the annotations in the @Controller, Swagger UI plugin will look at the Api interface, so the derived @Controller can be as simple as only defining the function parameters without any annotations.

Hope that helps!
